### ISSUE TITLE
UPnP / DLNA streaming on network receiver

### ISSUE THREAD

**Issue by sbaconnais** [theScrabi/NewPipe#105]
Chromecast is a closed source device with closed API.

But UPnP / DLNA are open protocols and many smart TVs implement it natively. It would be great to implement UPnP / DLNA stream to network compatible devices.

You can get source code from YAACC (GPL v3) here http://sourceforge.net/p/yaacc/code/ci/master/tree/

**Comment by medavox** [comment]
Consider it added to the queue.
If you feel like you can implement this yourself, feel free to have a go.

**Comment by sbaconnais** [comment]
Unfortunately i'm not an Android developer ... But i'm patient ;)

**Comment by medavox** [comment]
Fair enough :) I'll add it to the TODO list on the wiki.

**Comment by ildar** [comment]
I suggest the feature should be implemented on the Yaacc side: https://sourceforge.net/p/yaacc/tickets/117/

**Comment by sbaconnais** [comment]
Are you a Yaacc developer  ?

**Comment by ildar** [comment]
Nope

**Comment by sbaconnais** [comment]
Yaacc doesn't seem to be really active. And it's a bit outdated ...

I'm not even sure there are API that could be uses by NewPipe.

Perhaps including source code would be a better solution.

**Comment by pizzadude** [comment]
I know this is an old issue, but just in case anyone's interested, you can share a video from NewPipe to an application like BubbleUPNP or Allcast and it will work with the chromecast. But that's only if you're okay with installing proprietary apps..

(Those apps work with UPnP devices also)

**Comment by theScrabi** [comment]
I think if you use cromecast at the first place you also don't care about using non free software. However I'd like to support upnp but I don't know how, and to be honest, I have other problems to care about right first :/

**Comment by sbaconnais** [comment]
Here is an open source library for Upnp/Dlna : https://github.com/4thline/cling

**Comment by theScrabi** [comment]
All right. I can't do it right now, but if someone wants to implement this for NewPipe I'll add it.

**Comment by KeizerDev** [comment]
If I do a PR with Chromecast support, using https://github.com/googlecast/CastCompanionLibrary-android and the use of the default media controller would this be OK? I would like to support the Chromecast Audio for youtube because the official does not do it. 

If it's not OK i'll start working on my own client with similar support but I would rather like to do a collaboration.

**Comment by theScrabi** [comment]
Depends, is this library free software, or did we have to import binary blobs?

**Comment by KeizerDev** [comment]
@theScrabi Let's implement this: https://github.com/DroidsOnRoids/Casty

**Comment by theScrabi** [comment]
that went fast :P

**Comment by logos88** [comment]
I agree DLNA would be a great possibility. It would make possible to send videos to TVs with DLNA support, android TV boxes, etc. And it would help to spread this standard more, rather than the propietary chromecast

**Comment by wb9688** [comment]
@KeizerDev: Doesn't that depend on the proprietary Google Play Services?

---

I also agree that DLNA support would be great.

**Comment by theScrabi** [comment]
I guess I've asked this question before, but is there a library for java that can send DLNA streams?

**Comment by korelstar** [comment]
Here is an open-source library for DLNA on Android: https://github.com/4thline/cling

**Comment by wb9688** [comment]
@korelstar: We know about that library, but it's kinda bloated and relatively hard to use. Besides that it only support UPnP/DLNA and we also want to support Cast (v2) and probably other protocols. So we're creating our own library: [PipeCast](https://github.com/TeamNewPipe/PipeCast). It can already discover UPnP MediaRenderers, but not play to it. Besides that Android TV support has higher priority for me, so I'll finish that first.

**Comment by logos88** [comment]
https://github.com/ohmae/DmsExplorer

Maybe this could be useful

**Comment by logos88** [comment]
The yaacc developer has resumed his work!
https://github.com/tobexyz/yaacc-code

**Comment by IngwiePhoenix** [comment]
Did this ever land as a native feature? Didn't find it in the app.

I have an old Viera TV here that supposedly does DLNA and I would like to try that.

**Comment by BalooRJ1** [comment]
I too would like to know if Newpipe supports UPnP/DNLA as I have an Oppo BDP-103 that I would love to easily stream Youtube videos to.

**Comment by opusforlife2** [comment]
@BalooRJ1 It doesn't, but you can indirectly cast through VLC by using it as your external player.

**Comment by Atemu** [comment]
@opusforlife2 does that work for split a/v streams?

**Comment by opusforlife2** [comment]
I have no idea. Never used cast.

**Comment by BalooRJ1** [comment]
Thanks for the reply. I managed to find an app called BubbleUPnP which Newpipe can "share" videos to to app itself on Android that allows for it stream Newpipe videos just fine to any DNLA renderer like my Oppo BDP-103 for anyone interested. Works great, much better than casting to Kodi using the Kodi app. However BubbleUPnP is not completely open-source.

**Comment by opusforlife2** [comment]
What about Nova Video Player? (Video player for local/network content with subtitle/metadata download support)
https://f-droid.org/packages/org.courville.nova/

And other UPnP apps on F-Droid?

**Comment by BalooRJ1** [comment]
Thanks for that link, I'll give Nova or Yaaac a try. A prior search on F-Droid seemed to only bring up Plain UPnP and Droid UPnP, which did not work with my device.

**Comment by logos88** [comment]
Is [Pipecast](https://github.com/TeamNewPipe/PipeCast) dead?

**Comment by opusforlife2** [comment]
Pretty much, until a messiah contributor comes and revives it.
